cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LocalDatasetDeleterRunnable.java

```diff
@@ -0,0 +1,111 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime;
+
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.namespace.NamespaceAdmin;
+import co.cask.cdap.common.service.Retries;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.internal.app.store.RunRecordMeta;
+import co.cask.cdap.proto.DatasetSpecificationSummary;
+import co.cask.cdap.proto.NamespaceMeta;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.collect.ImmutableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class responsible for deleting the local datasets associated with the completed, failed, or killed workflow runs.
+ */
+public class LocalDatasetDeleterRunnable implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalDatasetDeleterRunnable.class);
+  private final NamespaceAdmin namespaceAdmin;
+  private final Store store;
+  private final DatasetFramework datasetFramework;
+  private final Map<String, String> properties;
+
+  public LocalDatasetDeleterRunnable(NamespaceAdmin namespaceAdmin, Store store, DatasetFramework datasetFramework) {
+    this.namespaceAdmin = namespaceAdmin;
+    this.store = store;
+    this.datasetFramework = datasetFramework;
+    this.properties = ImmutableMap.of(Constants.AppFabric.WORKFLOW_LOCAL_DATASET_PROPERTY, "true");
+  }
+
+  @Override
+  public void run() {
+    try {
+      List<NamespaceMeta> list = namespaceAdmin.list();
+      for (NamespaceMeta namespaceMeta : list) {
+        Collection<DatasetSpecificationSummary> specs
+          = datasetFramework.getInstances(namespaceMeta.getNamespaceId(), properties);
+
+        Set<String> activeRuns = getActiveRuns(namespaceMeta.getNamespaceId());
+        for (DatasetSpecificationSummary spec : specs) {
+          deleteLocalDataset(namespaceMeta.getName(), spec.getName(), activeRuns, spec.getProperties());
+        }
+      }
+    } catch (Throwable t) {
+      LOG.warn("Failed to delete the local datasets.", t);
+    }
+  }
+
+  private Set<String> getActiveRuns(NamespaceId namespaceId) {
+    Map<ProgramRunId, RunRecordMeta> activeRuns = store.getActiveRuns(namespaceId);
```

Comment:
Use the `Map<ProgramRunId, RunRecordMeta> getRuns(ProgramRunStatus status, Predicate<RunRecordMeta> filter);` method instead to fetch for the `STARTING`, `RUNNING` and `SUSPENDED` states and with workflow that has local dataset (based on the `Collection<DatasetSpecificationSummary>` above, assuming you turn that into a `Map<ProgramRunId, Collection<DataetSpecificationSummary>>`).